rfcs/text/0003_handler_interface.md

```diff
@@ -0,0 +1,299 @@
+Examples:
+
+```tsx
+// Task manager tasks
+taskManager.registerTaskDefinitions({
+  myTask: {
+    title: 'The task',
+    timeout: '5m',
+    createTaskRunner(context) {
+      return {
+        async run() {
+          const docs = await context.elasticsearch.search();
+          doSomethingWithDocs(docs);
+        }
+      }
+    }
+  }
+})
+
+// Application mount handlers
+application.registerApp({
+  id: 'myApp',
+  mount(context, domElement) {
+    ReactDOM.render(<MyApp overlaysService={context.overlays} />, domElement);
+    return () => ReactDOM.unmountComponentAtNode(domElement);
+  }
+});
+
+// Alerting
+alerting.registerType({
+  id: 'myAlert',
+  async execute(context, params, state) {
+    const indexPatterns = await context.savedObjects.find('indexPattern');
+    // use index pattern to search
+  }
+})
+```
+
+Without a formal definition, each handler interface varies slightly and
+different solutions are developed per handler for managing complexity and
+enabling extensibility.
+
+The official handler context convention seeks to address five key problems:
+
+1. Different services and plugins should be able to expose functionality that
+   is configured for the particular context where the handler is invoked, such
+   as a savedObject client in an alert handler already being configured to use
+   the appropriate API token.
+
+2. The service owner of a handler should not need to know about the services
+   or plugins that extend its handler context, such as the security plugin
+   providing a currentUser function to an HTTP router handler.
+
+3. Functionality in a handler should be "fixed" for the life of that
+   handler's context rather than changing configuration under the hood in
+   mid-execution. For example, while Elasticsearch clients can technically
+   be replaced throughout the course of the Kibana process, an HTTP route
+   handler should be able to depend on their being a consistent client for its
+   own shorter lifespan.
+
+4. Plugins should not need to pass down high level service contracts throughout
+   their business logic just so they can access them within the context of a
+   handler.
+
+5. Functionality provided by services should not be arbitrarily used in
+   unconstrained execution such as in the plugin lifecycle hooks. For example,
+   it's appropriate for an Elasticsearch client to throw an error if it's used
+   inside an API route and Elasticsearch isn't available, however it's not
+   appropriate for a plugin to throw an error in their start function if
+   Elasticsearch is not available. If the ES client was only made available
+   within the handler context and not to the plugin's start contract at large,
+   then this isn't an issue we'll encounter in the first place.
+
+# Detailed design
+
+There are two parts to this proposal. The first is the handler interface
+itself, and the second is the interface that a service owner implements to make
+their handlers extensible.
```

Comment:
Is it really making the handlers extensible, or would it be more accurate to say, extending the context that is passed into the handlers?

Reply:
I retract this, semantics are close enough.  Think I'm understanding.
 - Plugin developers can add new services.
 - Plugin developers can add handler functions to services.
 - Plugin developers can extend the context that is passed into handler functions for a given service.